content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1117 @@
+        * [`proc`](https://github.com/apple/darwin-xnu/blob/main/bsd/sys/proc_internal.h#L182) on XNU, the Apple iOS / macOS kernel
+    * All information about the process (from `ps`, `/proc`)
+
+----
+
+### Process Attributes - from Kernel Space
+
+The most important fields in `task_struct`:
+```c [2 | 4 | 6 | 8 - 9 | 11 - 13 | 15 - 16 | 18 - 19 | 21 | 23 | 25 - 26 | 28 - 29]
+struct task_struct {
+        unsigned int                    __state;
+
+        void                           *stack;
+
+        unsigned int                    flags;
+
+        int                             on_cpu;
+        int                             prio;
+
+        /* Scheduler information */
+        struct sched_entity             se;
+        const struct sched_class        *sched_class;
+
+        /* The VAS: memory mappings */
+        struct mm_struct                *mm;
+
+        int                             exit_state;
+        int                             exit_code;
+
+        pid_t                           pid;
+
+        struct task_struct __rcu        *parent;
+
+        /* Child processes */
+        struct list_head                children;
+
+        /* Open file information */
+        struct files_struct             *files;
+};
+```
+
+---
+
+### `fork`
+
+* Spawns child process
+* Child process starts execution right after the call to `fork`
+* Returns twice
+    * `0` in child process
+    * Child's PID in parent process
+* [Quiz](./quiz/process-creation.md)
+* Initially child and parent share the same PAS (Physical Address Space)
+* Child's writable pages are marked **copy-on-write**
+
+----
+
+### Copy-on-write
+
+When `fork` returns, the child duplicates the contents of the parent's page table.
+
+![Copy-on-write after fork()](./media/copy-on-write-initial.svg)
+
+----
+
+### Copy-on-write
+
+* Writable pages are **copied and modified only** when written: `demo/copy-on-write/copy_on_write_overhead.c`
+* [Quiz](./quiz/sections-always-shared.md)
+
+![Copy-on-write after Write](./media/copy-on-write-final.svg)
+
+----
+
+### Copy-on-write Overheard
+
+* Writes cause minor page faults
+
+```
+student@os:~/.../compute/lecture/demo/copy-on-write$ ./copy_on_write_overhead 
+ * Child process started
+ -- Press ENTER to continue ...
+Time for reading 131072 pages: 30 ms
+ * Child process read pages
+ -- Press ENTER to continue ...
+Time for writing to 131072 pages: 222 ms
+[...]
+```
+
+* The child process performs exactly 131072 **minor** page faults when writing data
+
+```
+student@os:~$ cat /proc/`pidof copy_on_write_overhead`/stat | cut -d ' ' -f 10  # before writing
+22
+student@os:~$ cat /proc/`pidof copy_on_write_overhead`/stat | cut -d ' ' -f 10  # after writing
+131094
+```
+
+---
+
+## Threads
+
+----
+
+## Threads - Backstory
+
+* You have an application that applies filters to a set of images
+* Processes need to share all images
+    * Difficult because the images (like all data) are marked copy-on-write
```

Comment:
```suggestion
* To optimize our application we use processes
    * Trade-off: copy-on-write
    * images (like all data) need to be shared between processes
```

Reply:
`copy-on-write` isn't a trade-off here. It's a downright drawback. I've reworded it differntly